src/Stratis.Bitcoin/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,318 @@
+                    linkedTokenSource.Cancel();
+                }
+
+                return Task.CompletedTask;
+            },
+            linkedTokenSource.Token,
+            repeatEvery: repeatEvery);
+        }
+
+        /// <inheritdoc />
+        public IAsyncQueue<T> CreateAsyncQueue<T>()
+        {
+            return new AsyncQueue<T>();
+        }
+
+        /// <inheritdoc />
+
+        public bool IsAsyncDelegateDequeuerRunning(IAsyncDelegate asyncDelegate)
+        {
+            lock (this.lockAsyncDelegates)
+            {
+                if (this.asyncDelegates.TryGetValue(asyncDelegate, out AsyncTaskInfo delegateInfo))
+                {
+                    // task in the dictionaries are either running or faulted so we just look for an asyncDelegate with a not faulted status.
+                    return delegateInfo.Status != TaskStatus.Faulted;
+                }
+
+                return false;
+            }
+        }
+
+        /// <inheritdoc />
+
+        public bool IsAsyncDelegateDequeuerRunning(string name)
+        {
+            lock (this.lockAsyncDelegates)
+            {
+                // task in the dictionaries are either running or faulted so we just look for an IAsyncDelegate with the given name and status not faulted.
+                return this.asyncDelegates.Values.Any(@delegate => @delegate.FriendlyName == name && @delegate.Status != TaskStatus.Faulted && @delegate.IsDelegateWorker == true);
+            }
+        }
+
+        /// <inheritdoc />
+
+        public bool IsAsyncLoopRunning(string name)
+        {
+            lock (this.lockAsyncDelegates)
+            {
+                // task in the dictionaries are either running or faulted so we just look for a dequeuer with the given name and status not faulted.
+                return this.asyncDelegates.Values.Any(@delegate => @delegate.FriendlyName == name && @delegate.Status != TaskStatus.Faulted && @delegate.IsLoop == true);
+            }
+        }
+
+        /// <inheritdoc />
+        [NoTrace]
+        public string GetStatistics(bool faultyOnly)
+        {
+            List<KeyValuePair<IAsyncDelegate, AsyncTaskInfo>> taskInformationsDump;
+            lock (this.lockAsyncDelegates)
+            {
+                // takes a snapshot of the informations.
+                taskInformationsDump = this.asyncDelegates.ToList();
+            }
+
+            int running = taskInformationsDump.Where(info => info.Value.IsRunning).Count();
+            int faulted = taskInformationsDump.Where(info => !info.Value.IsRunning).Count();
+
+            var sb = new StringBuilder();
+            sb.AppendLine();
+            sb.AppendLine($"====== Async loops ======   [Running: {running.ToString()}] [Faulted: {faulted.ToString()}]");
+
+            if (faultyOnly && faulted == 0)
+                return sb.ToString(); // If there are no faulty tasks and faultOnly is set to true, return just the header.
+
+            var data =
```

Comment:
Why linq :(

Reply:
a lot easier to deal with in a non performance critical part of code, are we against it?
I can refactor it on another PR if this is the only stopping point